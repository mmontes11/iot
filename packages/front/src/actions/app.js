import {
  TOGGLE_HAMBURGER_MENU,
  TOGGLE_MAP_DIALOG,
  TOGGLE_LANGUAGE_SELECTOR,
  SET_SHOW_ERROR,
} from "constants/actionTypes/app";

export const toggleHamburgerMenu = () => dispatch => {
  dispatch({ type: TOGGLE_HAMBURGER_MENU });
};

export const toggleMapDialog = () => dispatch => {
  dispatch({ type: TOGGLE_MAP_DIALOG });
};

export const toggleLanguageSelector = () => dispatch => {
  dispatch({ type: TOGGLE_LANGUAGE_SELECTOR });
};

export const setShowError = showError => dispatch => {
  dispatch({ type: SET_SHOW_ERROR, showError });
};
