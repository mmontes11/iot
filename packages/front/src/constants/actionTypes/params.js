export const PARAM_SELECT = "PARAM_SELECT";
export const PARAM_UPDATE = "PARAM_UPDATE";
export const PARAM_RESET = "PARAM_RESET";
export const PARAM_REQUEST = "PARAM_REQUEST";
export const PARAM_REQUEST_SUCCESS = "PARAM_REQUEST_SUCCESS";
export const PARAM_REQUEST_ERROR = "PARAM_REQUEST_ERROR";
export const PARAM_ITEMS_UPDATED = "PARAM_ITEMS_UPDATED";
export const PARAM_DISABLE = "PARAM_DISABLE";
