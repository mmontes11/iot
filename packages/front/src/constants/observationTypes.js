export const EVENT_TYPE = "event";
export const MEASUREMENT_TYPE = "measurement";
export const OBSERVATION_TYPES = [EVENT_TYPE, MEASUREMENT_TYPE];
