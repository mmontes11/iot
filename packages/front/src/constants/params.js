export const TYPE = "TYPE";
export const OBSERVATION = "OBSERVATION";
export const GROUPBY = "GROUPBY";
export const THING = "THING";
