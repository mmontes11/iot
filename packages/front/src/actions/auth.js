import {
  IS_AUTH,
  SET_USERNAME,
  SET_PASSWORD,
  LOGIN_REQUEST,
  LOGIN_REQUEST_SUCCESS,
  LOGIN_REQUEST_ERROR,
} from "constants/actionTypes/auth";
import { setShowError } from "actions/app";
import iotClient from "lib/iotClient";

export const isAuth = () => dispatch => {
  iotClient.authService
    .isAuth()
    .then(isAuthVal => dispatch({ type: IS_AUTH, isAuth: isAuthVal }))
    .catch(() => dispatch({ type: IS_AUTH, isAuth: false }));
};

export const setUsername = username => dispatch => {
  dispatch({ type: SET_USERNAME, username });
};

export const setPassword = password => dispatch => {
  dispatch({ type: SET_PASSWORD, password });
};

export const login = () => (dispatch, getState) => {
  dispatch({ type: LOGIN_REQUEST });
  const { username, password } = getState().auth;
  iotClient.authService.setCredentials(username, password);
  iotClient.authService
    .getToken()
    .then(response => {
      dispatch({
        type: LOGIN_REQUEST_SUCCESS,
        statusCode: response.statusCode,
        error: null,
      });
      isAuth()(dispatch);
    })
    .catch(error => {
      dispatch({
        type: LOGIN_REQUEST_ERROR,
        statusCode: error.statusCode,
        error,
      });
      setShowError(true)(dispatch);
    });
};

export const logout = () => dispatch => {
  iotClient.authService
    .logout()
    .then(() => {
      isAuth()(dispatch);
    })
    .catch(() => {
      isAuth()(dispatch);
    });
};
