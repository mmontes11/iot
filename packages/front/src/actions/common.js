import { RESET } from "constants/actionTypes/common";

export const reset = preserveError => dispatch => {
  dispatch({ type: RESET, preserveError });
};
