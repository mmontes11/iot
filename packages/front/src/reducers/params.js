import {
  PARAM_SELECT,
  PARAM_UPDATE,
  PARAM_RESET,
  PARAM_REQUEST,
  PARAM_REQUEST_SUCCESS,
  PARAM_REQUEST_ERROR,
  PARAM_ITEMS_UPDATED,
  PARAM_DISABLE,
} from "constants/actionTypes/params";
import { RESET } from "constants/actionTypes/common";
import deepEqual from "deep-equal";

export const initialState = {
  byId: {},
  allIds: [],
  isResetDisabled: true,
};

const paramInitialState = {
  items: [],
  isActive: false,
  isLoading: false,
  isDisabled: false,
  selectedItem: null,
};

const addParamIfNeeded = (state, paramId) => {
  if (state.allIds.includes(paramId)) {
    return state;
  }
  return {
    ...state,
    byId: {
      ...state.byId,
      [paramId]: paramInitialState,
    },
    allIds: [...state.allIds, paramId],
  };
};

const updateParam = (state, paramId, param) => ({
  ...state,
  byId: {
    ...state.byId,
    [paramId]: param,
  },
});

const allParamsInInitialState = state => Object.values(state.byId).every(param => deepEqual(param, paramInitialState));

export default (state = initialState, { type, param: paramId, selectedItem, items }) => {
  if (type === RESET) {
    return initialState;
  }
  if (!paramId) {
    return state;
  }
  let newState = { ...state };
  newState = addParamIfNeeded(newState, paramId);
  const param = newState.byId[paramId];
  switch (type) {
    case PARAM_SELECT:
      return updateParam(newState, paramId, { ...param, isActive: !param.isActive });
    case PARAM_UPDATE:
      return { ...updateParam(newState, paramId, { ...param, isActive: false, selectedItem }), isResetDisabled: false };
    case PARAM_RESET: {
      const newStateWithResetParam = updateParam(newState, paramId, paramInitialState);
      return { ...newStateWithResetParam, isResetDisabled: allParamsInInitialState(newStateWithResetParam) };
    }
    case PARAM_REQUEST:
      return updateParam(newState, paramId, { ...param, isLoading: true, isDisabled: true });
    case PARAM_REQUEST_SUCCESS:
      return updateParam(newState, paramId, { ...param, isLoading: false, isDisabled: false });
    case PARAM_REQUEST_ERROR:
      return updateParam(newState, paramId, { ...param, isLoading: false, isDisabled: false });
    case PARAM_ITEMS_UPDATED:
      return updateParam(newState, paramId, { ...param, items });
    case PARAM_DISABLE:
      return updateParam(newState, paramId, { ...param, isDisabled: true });
    default:
      return state;
  }
};

export const getParam = (state, paramName) => state.byId[paramName] || {};
