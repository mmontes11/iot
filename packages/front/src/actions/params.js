import {
  PARAM_SELECT,
  PARAM_UPDATE,
  PARAM_RESET,
  PARAM_REQUEST,
  PARAM_REQUEST_SUCCESS,
  PARAM_REQUEST_ERROR,
  PARAM_ITEMS_UPDATED,
  PARAM_DISABLE,
} from "constants/actionTypes/params";
import { EVENT_TYPE, MEASUREMENT_TYPE, OBSERVATION_TYPES } from "constants/observationTypes";
import { TYPE, OBSERVATION, GROUPBY, THING } from "constants/params";
import iotClient from "lib/iotClient";
import * as fromState from "reducers";
import { isEmpty } from "helpers/validation";

const requestObservations = (type, onStart, onSuccess, onError) => {
  if (type === EVENT_TYPE) {
    onStart();
    iotClient.eventService
      .getTypes()
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else if (type === MEASUREMENT_TYPE) {
    onStart();
    iotClient.measurementService
      .getTypes()
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  }
};

const requestObservationsDispatchingActions = (dispatch, type, cb) => {
  if (!type) {
    cb();
    return;
  }
  requestObservations(
    type,
    () => dispatch({ type: PARAM_REQUEST, param: OBSERVATION }),
    res => {
      dispatch({ type: PARAM_REQUEST_SUCCESS, param: OBSERVATION, statusCode: res.statusCode, error: null });
      dispatch({ type: PARAM_RESET, param: OBSERVATION });
      dispatch({ type: PARAM_ITEMS_UPDATED, param: OBSERVATION, items: res.body.types });
      cb();
    },
    error => {
      dispatch({ type: PARAM_REQUEST_ERROR, param: OBSERVATION, statusCode: error.statusCode, error });
      cb();
    },
  );
};

export const selectType = () => dispatch => {
  dispatch({ type: PARAM_ITEMS_UPDATED, param: TYPE, items: OBSERVATION_TYPES });
  dispatch({ type: PARAM_SELECT, param: TYPE });
};

export const updateType = type => dispatch => {
  dispatch({ type: PARAM_UPDATE, param: TYPE, selectedItem: type });
  requestObservationsDispatchingActions(dispatch, type, () => undefined);
};

export const selectObservation = () => (dispatch, getState) => {
  const state = getState();
  const observation = fromState.getParam(state, OBSERVATION);
  const dispatchSelectObservation = () => dispatch({ type: PARAM_SELECT, param: OBSERVATION });
  if (observation && !isEmpty(observation.items)) {
    dispatchSelectObservation();
  } else {
    const type = fromState.getParam(state, TYPE);
    if (type && type.selectedItem) {
      requestObservationsDispatchingActions(dispatch, type.selectedItem, dispatchSelectObservation);
    }
  }
};

export const updateObservation = observation => dispatch => {
  dispatch({ type: PARAM_UPDATE, param: OBSERVATION, selectedItem: observation });
};

export const selectGroupBy = () => (dispatch, getState) => {
  const state = getState();
  const groupBy = fromState.getParam(state, GROUPBY);
  if (groupBy && !isEmpty(groupBy.items)) {
    dispatch({ type: PARAM_SELECT, param: GROUPBY });
  } else {
    dispatch({ type: PARAM_REQUEST, param: GROUPBY });
    iotClient.timePeriodsService
      .getSupportedTimePeriods()
      .then(res => {
        dispatch({ type: PARAM_REQUEST_SUCCESS, param: GROUPBY, statusCode: res.statusCode, error: null });
        dispatch({ type: PARAM_ITEMS_UPDATED, param: GROUPBY, items: res.body.timePeriods });
        dispatch({ type: PARAM_SELECT, param: GROUPBY });
      })
      .catch(error => {
        dispatch({ type: PARAM_REQUEST_ERROR, param: GROUPBY, statusCode: error.statusCode, error });
      });
  }
};

export const updateGroupBy = groupBy => dispatch => {
  dispatch({ type: PARAM_UPDATE, param: GROUPBY, selectedItem: groupBy });
};

export const updateParams = params => dispatch => {
  const [firstParam, ...rest] = Object.keys(params);
  dispatch({ type: PARAM_UPDATE, param: firstParam, selectedItem: params[firstParam] });
  rest.forEach(key => {
    dispatch({ type: PARAM_DISABLE, param: key });
    dispatch({ type: PARAM_UPDATE, param: key, selectedItem: params[key] });
  });
};

export const selectThing = () => (dispatch, getState) => {
  const thing = fromState.getParam(getState(), THING);
  if (thing && !isEmpty(thing.items)) {
    dispatch({ type: PARAM_SELECT, param: THING });
  } else {
    dispatch({ type: PARAM_REQUEST, param: THING });
    iotClient.thingsService
      .getThings(true)
      .then(res => {
        dispatch({ type: PARAM_REQUEST_SUCCESS, param: THING, statusCode: res.statusCode, error: null });
        dispatch({ type: PARAM_ITEMS_UPDATED, param: THING, items: res.body.things });
        dispatch({ type: PARAM_SELECT, param: THING });
      })
      .catch(error => {
        dispatch({ type: PARAM_REQUEST_ERROR, param: THING, statusCode: error.statusCode, error });
      });
  }
};

export const updateThing = thingName => (dispatch, getState) => {
  dispatch({ type: PARAM_UPDATE, param: THING, selectedItem: thingName });
  const thingParam = fromState.getParam(getState(), THING);
  const thing = thingParam.items.find(item => item.name === thingName);
  dispatch({ type: PARAM_RESET, param: TYPE });
  dispatch({ type: PARAM_ITEMS_UPDATED, param: TYPE, items: thing.supportedObservationTypes.measurement });
};

export const selectMeasurementType = () => dispatch => {
  dispatch({ type: PARAM_SELECT, param: TYPE });
};

export const updateMeasurementType = type => dispatch => {
  dispatch({ type: PARAM_UPDATE, param: TYPE, selectedItem: type });
};
