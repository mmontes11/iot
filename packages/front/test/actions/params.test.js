import { selectType, updateType, selectObservation, updateObservation } from "actions/params";
import { EVENT_TYPE, MEASUREMENT_TYPE } from "constants/observationTypes";
import { OBSERVATION } from "constants/params";

describe("actions/params", () => {
  const testThunk = thunk => {
    const dispatch = jest.fn();
    thunk(dispatch);
    expect(thunk).toBeAThunk();
    expect(thunk).toMatchSnapshot();
  };
  it("dispatches a selectType action", () => {
    testThunk(selectType());
  });
  it("dispatches a updateType action", () => {
    let thunk = updateType(EVENT_TYPE);
    const dispatch = jest.fn();
    thunk(dispatch);
    expect(thunk).toBeAThunk();
    expect(thunk).toMatchSnapshot();
    thunk = updateType(MEASUREMENT_TYPE);
    thunk(dispatch);
    expect(thunk).toBeAThunk();
    expect(thunk).toMatchSnapshot();
  });
  it("dispatches a selectObservation action", () => {
    const dispatch = jest.fn();
    const getState = jest.fn(() => ({
      params: {
        byId: {
          OBSERVATION: {
            selectedItem: MEASUREMENT_TYPE,
          },
        },
        allIds: [OBSERVATION],
      },
    }));
    const thunk = selectObservation(dispatch, getState);
    expect(thunk).toBeAThunk();
    expect(thunk).toMatchSnapshot();
  });
  it("dispatches a updateObservation action", () => {
    testThunk(updateObservation());
  });
});
