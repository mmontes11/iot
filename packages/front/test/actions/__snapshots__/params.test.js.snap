// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions/params dispatches a selectObservation action 1`] = `[Function]`;

exports[`actions/params dispatches a selectType action 1`] = `[Function]`;

exports[`actions/params dispatches a updateObservation action 1`] = `[Function]`;

exports[`actions/params dispatches a updateType action 1`] = `[Function]`;

exports[`actions/params dispatches a updateType action 2`] = `[Function]`;
